examples/cuda/test_mm_opt.cpp

```diff
@@ -0,0 +1,401 @@
+        if (rc != HB_MC_SUCCESS) {
+                bsg_pr_err("Failed to allocate mat2 on device.\n");
+                return rc;
+        }
+
+        // Construct out
+        rc = hb_mc_device_malloc(&device, sizeof(hb_mc_device_tensor_t), &_out);
+        rc |= hb_mc_device_malloc(&device, out.dims * sizeof(uint32_t), &out.strides);
+        rc |= hb_mc_device_malloc(&device, out.dims * sizeof(uint32_t), &out.sizes);
+        rc |= hb_mc_device_malloc(&device, out.N    * sizeof(*Hout.data), &out.data);
+        if (rc != HB_MC_SUCCESS) {
+                bsg_pr_err("Failed to allocate out on device.\n");
+                return rc;
+        }
+
+        // Initialize RNG
+        // Initialize the random number generators
+        std::numeric_limits<float> lim;
+
+        std::mt19937 generator;
+        generator.seed(42);
+        std::uniform_real_distribution<float> distribution(-1,1);
+
+        //************************************************************
+        // Allocate memory on the host for mat1, mat2, and result
+        // and initialize the first three with random values.
+        //************************************************************
+        Hmat1.data = new float[Hmat1.N];
+        Hmat1.sizes = new uint32_t[Hmat1.dims];
+        Hmat1.sizes[0] = M;
+        Hmat1.sizes[1] = N;
+        Hmat1.strides = new uint32_t[Hmat1.dims];
+        Hmat1.strides[0] = Hmat1.sizes[1];
+        Hmat1.strides[1] = 1;
+        for (uint64_t i = 0; i < Hmat1.N; ++i) {
+                Hmat1.data[i] = distribution(generator);
+        }
+
+        Hmat2.data = new float[Hmat2.N];
+        Hmat2.sizes = new uint32_t[Hmat2.dims];
+        Hmat2.sizes[0] = N;
+        Hmat2.sizes[1] = P;
+        Hmat2.strides = new uint32_t[Hmat2.dims];
+        Hmat2.strides[0] = Hmat2.sizes[1];
+        Hmat2.strides[1] = 1;
+        for (uint64_t i = 0; i < Hmat2.N; ++i) {
+                Hmat2.data[i] = distribution(generator);
+        }
+
+        Hout.data = new float[Hout.N];
+        Hout.sizes = new uint32_t[Hout.dims];
+        Hout.sizes[0] = M;
+        Hout.sizes[1] = P;
+        Hout.strides = new uint32_t[Hout.dims];
+        Hout.strides[0] = Hout.sizes[1];
+        Hout.strides[1] = 1;
+
+        Hresult.data = new float[Hresult.N];
+        Hresult.sizes = new uint32_t[Hresult.dims];
+        Hresult.sizes[0] = M;
+        Hresult.sizes[1] = P;
+        Hresult.strides = new uint32_t[Hresult.dims];
+        Hresult.strides[0] = Hresult.sizes[1];
+        Hresult.strides[1] = 1;
+
+        //************************************************************
+        // Copy mat1 and mat2, from host onto device
+        //************************************************************
+        void *dst, *src;
+
+        // Copy mat1
+        bsg_pr_info("Copying mat1\n");
+        dst = (void *) ((intptr_t) _mat1);
+        src = (void *) &mat1;
+        rc = hb_mc_device_memcpy (&device, dst, src, sizeof(mat1), HB_MC_MEMCPY_TO_DEVICE);
```

Comment:
This is a suggestion more than a change request. CUDA lite has a macro that you can wrap around API calls to do error checking and reporting for you. I think it’s BSG_CUDA_CALL - defined in bsg_manycore_cuda.h. It will reduce the number of LOC and it will also print out the line of code that errored out for you.